tc/tc_util.c

```diff
@@ -192,6 +192,86 @@ static const struct rate_suffix {
 };
 
 
+int get_percent_rate(unsigned int *rate, const char *str, char *dev)
+{
+	long value;
+	char percent[strlen(dev) - 1], r_str[20];
+	const char *pos;
+	int ret;
+	double d, r;
+
+	if (read_prop(dev, "speed", &value))
+		return -1;
+
+	pos = str + strlen(str) -1;
+	if (*pos != '%') {
+		fprintf(stderr, "Specified rate value is malformed \n");
+		return -1;
+	}
+
+	strncpy(percent, str, strlen(str) -1);
+	ret = sscanf(percent, "%lf", &d);
+	if (ret != 1) {
+		fprintf(stderr, "Specified rate value is malformed \n");
+		return -1;
+	}
+
+	if (d > 100.0 || d < 0.0) {
+		fprintf(stderr, "Invalid rate specified \n");
+		return -1;
+	}
+
+	r = (d*value) / 100.0;
+
+	ret = snprintf(r_str, 20, "%f", r);
+	if (ret <= 0 || ret >= 20) {
+		fprintf(stderr, "Invalid rate specified \n");
+		return -1;
+	}
+
+	return get_rate(rate, r_str);
+}
+
+int get_percent_rate64(__u64 *rate, const char *str, char *dev)
+{
+	long value;
+	char percent[strlen(dev) - 1], r_str[20];
+	const char *pos;
+	int ret;
+	double d, r;
+
+	if (read_prop(dev, "speed", &value))
+		return -1;
+
+	pos = str + strlen(str) -1;
+	if (*pos != '%') {
+		fprintf(stderr, "Specified rate value is malformed \n");
+		return -1;
+	}
+
+	strncpy(percent, str, strlen(str) -1);
+	ret = sscanf(percent, "%lf", &d);
+	if (ret != 1) {
+		fprintf(stderr, "Specified rate value is malformed \n");
+		return -1;
+	}
+
+	if (d > 100.0 || d < 0.0) {
+		fprintf(stderr, "Invalid rate specified \n");
+		return -1;
+	}
+
+	r = (d*value) / 100.0;
+	ret = snprintf(r_str, 20, "%f", r);
+	if (ret <= 0 || ret >= 20) {
+		fprintf(stderr, "Invalid rate specified \n");
+		return -1;
+	}
+
+	return get_rate64(rate, r_str);
+
+}
```

Comment:
This function almost completely duplicates the code of `get_percent_rate()`, so you should find a way of putting the code that calculates the bandwidth rate in its own function.

Reply:
Done. But now the problem is that the `get_percent_rate` and `get_percent_rate64` only look like wrappers calling two functions inside them and that's it. Is that ok?

Reply:
Also another thing I pointed out is that the rate parse functions `get_rate` and `get_rate64` do not explicity state the error in the given command, they just `return -1` and let the `explain` functions in the qdisc file handle the explanation of what was wrong. We, as of right now, go against that norm and explicitly state what went wrong inside the parse functions. Should we change this (and modify the `explain` functions in the qdisc files), what do you think?

Reply:
I'll need to look at what you ultimately did to factor this code out, but it should be OK that these functions now look like wrappers.

As for error logging, I think what we're doing is OK. I see other instances of using `fprintf()` individually (I'm looking in `tc/q_cbq.c` specifically). It looks like functions like `explain()` and `explain1()` suit more narrower purposes.